File: MyMechanicAgenda/app/src/main/java/com/santosdaniel/mymechanicagenda/presenter/network/LocalImageLoader.java
package com.santosdaniel.mymechanicagenda.presenter.network;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Matrix;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.provider.MediaStore;
import android.support.annotation.NonNull;
import android.util.Log;
import android.widget.ImageView;
import android.widget.LinearLayout;

import static com.android.volley.toolbox.ImageLoader.ImageCache;

/**
 * Loads one image locally
 */
public class LocalImageLoader {

    private final Context context;
    private final ImageCache imageCache;

    /**
     * Constructs a new ImageLoader.
     *
     * @param imageCache The cache to use as an L1 cache.
     */
    public LocalImageLoader(@NonNull Context context, @NonNull ImageCache imageCache) {
        this.context = context;
        this.imageCache = imageCache;
    }

    private int dpToPx(int dp) {
        float density = context.getResources().getDisplayMetrics().density;
        return Math.round((float) dp * density);
    }

    private void scaleImage(ImageView imageView) {

        Drawable drawing = imageView.getDrawable();
        if (drawing == null) {
        }
        Bitmap bitmap = ((BitmapDrawable) drawing).getBitmap();

        int width = bitmap.getWidth();
        int height = bitmap.getHeight();
        int bounding = dpToPx(50);

        float xScale = ((float) bounding) / width;
        float yScale = ((float) bounding) / height;
        float scale = (xScale <= yScale) ? xScale : yScale;
        Matrix matrix = new Matrix();
        matrix.postScale(scale, scale);

        Bitmap scaledBitmap = Bitmap.createBitmap(bitmap, 0, 0, width, height,
                matrix, true);
        width = scaledBitmap.getWidth(); // re-use
        height = scaledBitmap.getHeight(); // re-use
        BitmapDrawable result = new BitmapDrawable(scaledBitmap);

        imageView.setImageDrawable(result);

        LinearLayout.LayoutParams params = (LinearLayout.LayoutParams) imageView
                .getLayoutParams();
        params.width = width;
        params.height = height;
        imageView.setLayoutParams(params);

    }

    /**
     *
     *
     * @param imageUri
     * @param thumbnail
     */
    public void load(String imageUri, ImageView thumbnail) {
        Bitmap bitmap = imageCache.getBitmap(imageUri);
        if(bitmap == null) {
            try {
                Uri uri = Uri.parse(imageUri);
                bitmap = MediaStore.Images.Media.getBitmap(
                        context.getContentResolver(), uri);
                Log.d("Tag", bitmap.getWidth() + "");
                thumbnail.setImageBitmap(bitmap);
                scaleImage(thumbnail);
            } catch (Exception e) {
            }
        } else {
            thumbnail.setImageBitmap(bitmap);
        }
    }
}

File: MyMechanicAgenda/app/src/main/java/com/santosdaniel/mymechanicagenda/presenter/network/LocalImageLoader.kt
package com.santosdaniel.mymechanicagenda.presenter.network

import android.content.Context
import android.graphics.Bitmap
import android.net.Uri
import android.provider.MediaStore
import android.support.v4.content.FileProvider
import android.util.Log
import android.widget.ImageView
import com.android.volley.toolbox.ImageLoader.ImageCache
import com.santosdaniel.mymechanicagenda.helper.TakePictureHelper
import java.io.File
import java.io.FileNotFoundException

/**
 * Loads one image locally
 */
class LocalImageLoader
/**
 * Constructs a new ImageLoader.
 *
 * @param imageCache The cache to use as an L1 cache.
 */
(private val context: Context, private val imageCache: ImageCache) {


    private fun load(path: String, uri: Uri, thumbnail: ImageView) {
        val bitmap: Bitmap? = MediaStore.Images.Media.getBitmap(
                context.contentResolver, uri)
        thumbnail.setImageBitmap(bitmap)
        if (bitmap != null) {
            imageCache.putBitmap(path, bitmap)
        }
    }

    /**
     *
     *
     * @param imageUri
     * @param thumbnail
     */
    fun load(imageUri: String, thumbnail: ImageView): Boolean {
        val bitmap: Bitmap? = imageCache.getBitmap(imageUri)
        if (bitmap == null) {
            try {
                val uri = Uri.parse(imageUri)
                load(imageUri, uri, thumbnail)
                return true
            } catch (fNotFound: FileNotFoundException) {
                try {
                    val photoFile = File(imageUri)
                    val photoURI = FileProvider.getUriForFile(context,
                            TakePictureHelper.FILE_PROVIDER_AUTHORITY,
                            photoFile)
                    load(imageUri, photoURI, thumbnail)
                    return true
                } catch (e: Exception) {
                    Log.e(TAG, "Impossible to load image", e)
                    return false
                }
            }
        } else {
            //Was possible to get the image from the cache so is going to used it
            thumbnail.setImageBitmap(bitmap)
            return true
        }
    }

    companion object {

        /**
         * Maximum number of entries in the cache
         */
        private const val TAG = "LocalImageLoader"
    }
}
